File: android/src/main/java/cn/pocket/agora/AgoraModel.java
package cn.pocket.agora;


import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.RCTNativeAppEventEmitter;

import io.agora.AgoraAPIOnlySignal;
import io.agora.AgoraAPI;

/**
 * Created by jessica on 16/12/15.
 */

public class AgoraModel extends ReactContextBaseJavaModule {

    private static AgoraAPIOnlySignal agoraAPI;
    final String EVENT_NAME = "Agora_Resp";

    @Override
    public String getName() {
        return "AgoraModel";
    }

    public AgoraModel(ReactApplicationContext context){
        super(context);
        agoraAPI = AgoraAPIOnlySignal.getInstance(context, AgoraViewManager.getAppId());
        agoraAPI.callbackSet(new AgoraAPI.CallBack() {

            private void sendMessage(WritableMap map) {
                RCTNativeAppEventEmitter emitter = getReactApplicationContext().getJSModule(RCTNativeAppEventEmitter.class);
                emitter.emit(EVENT_NAME, map);
            }
            @Override
            public void onLoginSuccess(int uid, int fd) {
                super.onLoginSuccess(uid, fd);
                System.out.println("onLoginSuccess========" + uid + "===" + fd);
                WritableMap map = Arguments.createMap();
                map.putString("actionType", "loginSuccess");
                map.putString("uid", "" + uid);
                map.putString("fd", "" + fd);
                sendMessage(map);
            }

            @Override
            public void onLogout(int ecode) {
                super.onLogout(ecode);
                System.out.println("onLogout========" + ecode);
                WritableMap map = Arguments.createMap();
                map.putString("actionType", "logout");
                map.putString("ecode", "" + ecode);
                sendMessage(map);
            }

            @Override
            public void onLoginFailed(int ecode) {
                super.onLoginFailed(ecode);
                System.out.println("onLoginFailed========" + ecode);
                WritableMap map = Arguments.createMap();
                map.putString("actionType", "loginFailed");
                map.putString("ecode", "" + ecode);
                sendMessage(map);
            }

            @Override
            public void onChannelJoined(String channelID) {
                super.onChannelJoined(channelID);
                System.out.println("onChannelJoined========" + channelID);
                WritableMap map = Arguments.createMap();
                map.putString("actionType", "channelJoined");
                map.putString("channel", channelID);
                sendMessage(map);
            }

            @Override
            public void onChannelJoinFailed(String channelID, int ecode) {
                super.onChannelJoined(channelID);
                System.out.println("onChannelJoined========" + channelID);
                WritableMap map = Arguments.createMap();
                map.putString("actionType", "channelJoinFailed");
                map.putString("channel", channelID);
                map.putString("ecode", "" + ecode);
                sendMessage(map);
            }

            @Override
            public void onChannelLeaved(String channelID, int ecode) {
                super.onChannelJoined(channelID);
                System.out.println("onChannelLeaved========" + channelID);
                WritableMap map = Arguments.createMap();
                map.putString("actionType", "channelLeaved");
                map.putString("channel", channelID);
                map.putString("ecode", "" + ecode);
                sendMessage(map);
            }

            @Override//收到呼叫邀请回调
            public void onInviteReceived(String channelID, String account, int uid, String extra) {
                super.onInviteReceived(channelID, account, uid, extra);
                System.out.println("onInviteReceived========" + channelID);
                WritableMap map = Arguments.createMap();
                map.putString("actionType", "inviteReceived");
                map.putString("channel", channelID);
                map.putString("account", account);
                map.putString("uid", "" + uid);
                map.putString("extra", extra);
                sendMessage(map);
            }

            @Override//远端已收到呼叫回调
            public void onInviteReceivedByPeer(String channelID, String account, int uid) {
                super.onInviteReceivedByPeer(channelID, account, uid);
                System.out.println("onInviteReceivedByPeer========" + channelID);
                WritableMap map = Arguments.createMap();
                map.putString("actionType", "inviteReceivedByPeer");
                map.putString("channel", channelID);
                map.putString("account", account);
                map.putString("uid", "" + uid);
                sendMessage(map);
            }

            @Override
            public void onInviteAcceptedByPeer(String channelID, String account, int uid) {
                super.onInviteAcceptedByPeer(channelID, account, uid);
                System.out.println("onInviteAcceptedByPeer========" + channelID);
                WritableMap map = Arguments.createMap();
                map.putString("actionType", "inviteAcceptedByPeer");
                map.putString("channel", channelID);
                map.putString("account", account);
                map.putString("uid", "" + uid);
                sendMessage(map);
            }

            @Override
            public void onInviteRefusedByPeer(String channelID, String account, int uid) {
                super.onInviteAcceptedByPeer(channelID, account, uid);
                System.out.println("onInviteAcceptedByPeer========" + channelID);
                WritableMap map = Arguments.createMap();
                map.putString("actionType", "inviteRefusedByPeer");
                map.putString("channel", channelID);
                map.putString("account", account);
                map.putString("uid", "" + uid);
                sendMessage(map);
            }

            @Override
            public void onInviteEndByPeer(String channelID, String account, int uid) {
                super.onInviteEndByPeer(channelID, account, uid);
                System.out.println("onInviteEndByPeer========" + channelID);
                WritableMap map = Arguments.createMap();
                map.putString("actionType", "inviteEndByPeer");
                map.putString("channel", channelID);
                map.putString("account", account);
                map.putString("uid", "" + uid);
                sendMessage(map);
            }

            @Override
            public void onInviteEndByMyself(String channelID, String account, int uid) {
                super.onInviteEndByMyself(channelID, account, uid);
                System.out.println("onInviteEndByMyself========" + channelID);
                WritableMap map = Arguments.createMap();
                map.putString("actionType", "inviteEndByMyself");
                map.putString("channel", channelID);
                map.putString("account", account);
                map.putString("uid", "" + uid);
                sendMessage(map);
            }

            @Override
            public void onInviteFailed(String channelID, String account, int uid, int ecode) {
                super.onInviteFailed(channelID, account, uid, ecode);
                System.out.println("onInviteFailed========" + channelID);
                WritableMap map = Arguments.createMap();
                map.putString("actionType", "inviteFailed");
                map.putString("channel", channelID);
                map.putString("account", account);
                map.putString("uid", "" + uid);
                map.putString("ecode", "" + ecode);
                sendMessage(map);
            }

            @Override
            public void onError(String name, int ecode, String desc) {
                super.onError(name, ecode, desc);
                System.out.println("onError========" + name + "===" + ecode + "===" + desc);
                WritableMap map = Arguments.createMap();
                map.putString("actionType", "error");
                map.putString("name", name);
                map.putString("ecode", "" + ecode);
                map.putString("desc", desc);
                sendMessage(map);
            }
        });
    }

    @ReactMethod
    public void login(String signalKey, String account) {
        if(agoraAPI.isOnline() == 1) {
            agoraAPI.logout();
        }
        agoraAPI.login(AgoraViewManager.getAppId(), account, signalKey, 0, null);
    }

    @ReactMethod
    public void call(String channel, String account) {
        agoraAPI.channelInviteUser(channel, account, 0);
    }

    @ReactMethod
    public void accept(String channel, String account) {
        agoraAPI.channelInviteAccept(channel, account, 0);
    }

    @ReactMethod
    public void refuse(String channel, String account) {
        agoraAPI.channelInviteRefuse(channel, account, 0);
    }

    @ReactMethod
    public void end(String channel, String account) {
        agoraAPI.channelInviteEnd(channel, account, 0);
    }

    @ReactMethod
    public void joinChannel(String channelKey, String channelName) {
        AgoraView.getRtcEngine().joinChannel(channelKey, channelName, "", 0);
    }

    @ReactMethod
    public void leaveChannel(String channel) {
        this.leaveWithChannel(channel);
    }

    @ReactMethod
    public void logout() {
        agoraAPI.logout();
    }

    private void leaveWithChannel(String channel) {
        AgoraView.getRtcEngine().setupLocalVideo(null);
        AgoraView.getRtcEngine().setupRemoteVideo(null);
        AgoraView.getRtcEngine().leaveChannel();
    }

    public static AgoraAPIOnlySignal getAgoraAPI() {
        return agoraAPI;
    }
}

File: ios/AgoraManagerIOS/AgoraModel.m
//
//  RCTAgoraModel.m
//  agora
//
//  Created by jessica on 16/12/8.
//  Copyright © 2016年 Facebook. All rights reserved.
//

#import "AgoraModel.h"
#import "RCTConvert.h"
#import "RCTAgoraView.h"
#import "agorasdk.h"
#import "RCTAgoraManager.h"

static AgoraAPI* agoraApi;

@implementation AgoraModel
//  NSString* _signalKey;
//  NSString* _channelKey;
//  NSString* _channel;
//  NSString* _account;

@synthesize bridge = _bridge;
RCT_EXPORT_MODULE()

-(NSArray<NSString *> *) supportedEvents {
  return @[@"Agora_Resp"];
}

+(AgoraAPI*) agoraAPI {
  return agoraApi;
}

- (instancetype)init
{
  self = [super init];
  NSLog(@"AgoraModel===, init");
  if(agoraApi == nil) {
    agoraApi = [AgoraAPI getInstanceWithoutMedia:[RCTAgoraManager appId]];
    agoraApi.onLoginSuccess         = ^(uint32_t uid, int fd) {
      NSLog(@"onLoginSuccess===登陆成功, %u, %d", uid, fd);
      NSMutableDictionary *body = @{}.mutableCopy;
      body[@"actionType"] = @"loginSuccess";
      body[@"uid"] = [NSString stringWithFormat:@"%u", uid];
      body[@"fd"] = [NSString stringWithFormat:@"%d", fd];
      [self sendEventWithName:@"Agora_Resp" body:body];
    };
    agoraApi.onLogout               = ^(AgoraEcode ecode) {
      NSLog(@"onLogout===登出, %lu", ecode);
      NSMutableDictionary *body = @{}.mutableCopy;
      body[@"actionType"] = @"logout";
      [self sendEventWithName:@"Agora_Resp" body:body];
    };
    agoraApi.onLoginFailed          = ^(AgoraEcode ecode) {
      NSLog(@"onLoginFailed===登录失败, %lu", ecode);
      NSMutableDictionary *body = @{}.mutableCopy;
      body[@"actionType"] = @"loginFailed";
      body[@"ecode"] = [NSString stringWithFormat:@"%lu", ecode];
      [self sendEventWithName:@"Agora_Resp" body:body];
    };
    agoraApi.onChannelJoined        = ^(NSString* channel) {
      NSLog(@"onChannelJoined===加入频道, %@", channel);
      NSMutableDictionary *body = @{}.mutableCopy;
      body[@"actionType"] = @"channelJoined";
      body[@"channel"] = channel;
      [self sendEventWithName:@"Agora_Resp" body:body];
    };
    agoraApi.onChannelJoinFailed    = ^(NSString* channel,AgoraEcode ecode) {
      NSLog(@"onChannelJoined===加入频道失败, %@, %lu", channel, ecode);
      NSMutableDictionary *body = @{}.mutableCopy;
      body[@"actionType"] = @"channelJoinFailed";
      body[@"channel"] = channel;
      body[@"ecode"] = [NSString stringWithFormat:@"%lu", ecode];
      [self sendEventWithName:@"Agora_Resp" body:body];
    };
    agoraApi.onChannelLeaved        = ^(NSString* name, AgoraEcode ecode) {
      NSLog(@"onChannelLeaved===离开频道, %@ %lu", name, ecode);
      NSMutableDictionary *body = @{}.mutableCopy;
      body[@"actionType"] = @"channelLeaved";
      body[@"name"] = name;
      body[@"ecode"] = [NSString stringWithFormat:@"%lu", ecode];
      [self sendEventWithName:@"Agora_Resp" body:body];
    };
    agoraApi.onInviteReceived       = ^(NSString* channel, NSString *name, uint32_t uid) {
      NSLog(@"onInviteReceived===收到呼叫,%@ %@ %u",channel, name, uid);
      NSMutableDictionary *body = @{}.mutableCopy;
      body[@"actionType"] = @"inviteReceived";
      body[@"channel"] = channel;
      body[@"name"] = name;
      body[@"uid"] = [NSString stringWithFormat:@"%u", uid];
      [self sendEventWithName:@"Agora_Resp" body:body];
      
    };
    agoraApi.onInviteReceivedByPeer = ^(NSString* channel, NSString *name, uint32_t uid) {
      NSLog(@"onInviteReceivedByPeer===对方收到呼叫,%@ %@ %u",channel, name, uid);
      NSMutableDictionary *body = @{}.mutableCopy;
      body[@"actionType"] = @"inviteReceivedByPeer";
      body[@"channel"] = channel;
      body[@"name"] = name;
      body[@"uid"] = [NSString stringWithFormat:@"%u", uid];
      [self sendEventWithName:@"Agora_Resp" body:body];
    };
    agoraApi.onInviteAcceptedByPeer = ^(NSString* channel, NSString *name, uint32_t uid) {
      NSLog(@"onInviteAcceptedByPeer===对方接受呼叫,%@ %@ %u",channel, name, uid);
      NSMutableDictionary *body = @{}.mutableCopy;
      body[@"actionType"] = @"inviteAcceptedByPeer";
      body[@"channel"] = channel;
      body[@"name"] = name;
      body[@"uid"] = [NSString stringWithFormat:@"%u", uid];
      [self sendEventWithName:@"Agora_Resp" body:body];
    };
    agoraApi.onInviteRefusedByPeer  = ^(NSString* channel, NSString *name, uint32_t uid) {
      NSLog(@"onInviteRefusedByPeer===对方拒绝呼叫,%@ %@ %u",channel, name, uid);
      NSMutableDictionary *body = @{}.mutableCopy;
      body[@"actionType"] = @"inviteRefusedByPeer";
      body[@"channel"] = channel;
      body[@"name"] = name;
      body[@"uid"] = [NSString stringWithFormat:@"%u", uid];
      [self sendEventWithName:@"Agora_Resp" body:body];
    };
    agoraApi.onInviteEndByPeer      = ^(NSString* channel, NSString *name, uint32_t uid) {
      NSLog(@"onInviteEndByPeer===对方接通后结束呼叫,%@ %@ %u",channel, name, uid);
      NSMutableDictionary *body = @{}.mutableCopy;
      body[@"actionType"] = @"inviteEndByPeer";
      body[@"channel"] = channel;
      body[@"name"] = name;
      body[@"uid"] = [NSString stringWithFormat:@"%u", uid];
      [self sendEventWithName:@"Agora_Resp" body:body];

      //      [self leaveChannel];
    };
    agoraApi.onInviteEndByMyself    = ^(NSString* channel, NSString *name, uint32_t uid) {
      //      [self log:[NSString stringWithFormat:@"我方接通后关闭呼叫 %@ %@ %u", channel, name, uid]];
      NSLog(@"onInviteEndByMyself===我方接通后关闭呼叫,%@ %@ %u", channel, name, uid);
      NSMutableDictionary *body = @{}.mutableCopy;
      body[@"actionType"] = @"inviteEndByMyself";
      body[@"channel"] = channel;
      body[@"name"] = name;
      body[@"uid"] = [NSString stringWithFormat:@"%u", uid];
      [self sendEventWithName:@"Agora_Resp" body:body];
    };
    agoraApi.onInviteFailed         = ^(NSString* channel, NSString* account, uint32_t uid, AgoraEcode ecode) {
      NSLog(@"onInviteFailed===呼叫失败,%@ %@ %u %lu", channel, account, uid, ecode);
      NSMutableDictionary *body = @{}.mutableCopy;
      body[@"actionType"] = @"inviteFailed";
      body[@"channel"] = channel;
      body[@"account"] = account;
      body[@"uid"] = [NSString stringWithFormat:@"%u", uid];
      body[@"ecode"] = [NSString stringWithFormat:@"%lu", ecode];
      [self sendEventWithName:@"Agora_Resp" body:body];
    };
    agoraApi.onChannelUserLeaved    = ^(NSString* name, uint32_t uid){
//      [self log:[NSString stringWithFormat:@"用户离开频道 %@ %u", name, uid]];
    };
    agoraApi.onLog                  = ^(NSString *txt) {
//      [self log:[NSString stringWithFormat:@"onLog %@", txt]];
    };
    agoraApi.onError                = ^(NSString* name,AgoraEcode ecode,NSString* desc) {
      NSLog(@"onError===出现错误,%@ %lu %@", name, ecode, desc);
      NSMutableDictionary *body = @{}.mutableCopy;
      body[@"actionType"] = @"error";
      body[@"name"] = name;
      body[@"ecode"] = [NSString stringWithFormat:@"%lu", ecode];
      body[@"desc"] = desc;
      [self sendEventWithName:@"Agora_Resp" body:body];
    };
;
  }
  return self;
}

-(void) leaveWithChannel: (NSString*) channel {
    NSLog(@"leaveChannel===, %@", channel);
  
  [[RCTAgoraView agoraKit] setupLocalVideo:nil];
  [[RCTAgoraView agoraKit] setupRemoteVideo:nil];
  [[RCTAgoraView agoraKit] leaveChannel:^(AgoraRtcStats *stat) {
    NSLog(@"leaveChannel===completed, %@", stat);
    [agoraApi logout];
    [[RCTAgoraView agoraKit] stopPreview];
  }];
}

RCT_EXPORT_METHOD(login: (NSString*) signalKey account:(NSString*) account) {
  NSLog(@"login===, %@, %@", signalKey, account);
  if([agoraApi isOnline]) {
    [agoraApi logout];
  }
  [agoraApi login:[RCTAgoraManager appId] account:account token:signalKey uid:0 deviceID:nil];
}

RCT_EXPORT_METHOD(call:(NSString*) channel account:(NSString*) account) {
  NSLog(@"call=== %@, %@", channel, account);
  [agoraApi channelInviteUser:channel account:account uid:0];
}

RCT_EXPORT_METHOD(accept:(NSString*) channel account:(NSString*) account) {
  NSLog(@"accept===, %@, %@", channel, account);
  [agoraApi channelInviteAccept:channel account:account uid:0];
}

RCT_EXPORT_METHOD(refuse:(NSString *) channel account:(NSString*) account) {
  NSLog(@"refuse===, %@, %@", channel, account);
  [agoraApi channelInviteRefuse:channel account:account uid:0];
}

RCT_EXPORT_METHOD(end:(NSString *) channel account:(NSString*) account) {
  NSLog(@"end===, %@, %@", channel, account);
  [agoraApi channelInviteEnd:channel account:account uid:0];
//  [self leaveWithChannel:channel];
}

RCT_EXPORT_METHOD(joinChannel:(NSString*)key channelName:(NSString*) channelName) {
  NSLog(@"joinChannel===, %@,%@", key, channelName);
  [[RCTAgoraView agoraKit] joinChannelByKey:key channelName:channelName info:nil uid:0 joinSuccess:^(NSString *channel, NSUInteger uid, NSInteger elapsed) {
    NSLog(@"joinChannel=== %@ sucess", channelName);
  }];
}

RCT_EXPORT_METHOD(leaveChannel:(NSString*) channel) {
  [self leaveWithChannel:channel];
}

RCT_EXPORT_METHOD(logout) {
  NSLog(@"logout===");
  [agoraApi logout];
}

//RCT_EXPORT_METHOD(isOnline: (RCTResponseSenderBlock)callback) {
//  callback(@[@([agoraApi isOnline])]);
//}

@end
